### ISSUE TITLE
explicit_auto_deref: introduces type error

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9901]
### Summary

.

### Lint Name

explicit_auto_deref

### Reproducer

I tried this code:

```rust
// check-pass

use std::ops::Deref;

struct Data {
    boxed: Box<&'static i32>
}

impl Data {
    fn use_data(&self, user: impl for <'a> FnOnce(<Box<&'a i32> as Deref>::Target)) {
        user(*self.boxed)
    }
}

fn main() {}
```

I saw this happen:
` cargo clippy --fix -- -Aclippy::all -Wclippy::explicit_auto_deref`
```
warning: failed to automatically apply fixes suggested by rustc to crate `clpy`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
 --> src/main.rs:9:14
  |
9 |         user(self.boxed)
  |         ---- ^^^^^^^^^^
  |         |    |
  |         |    expected `&i32`, found struct `std::boxed::Box`
  |         |    help: consider borrowing here: `&self.boxed`
  |         arguments to this function are incorrect
  |
  = note: expected reference `&i32`
                found struct `std::boxed::Box<&'static i32>`
note: type parameter defined here
 --> src/main.rs:8:30
  |
8 |     fn use_data(&self, user: impl for <'a> FnOnce(<Box<&'a i32> as Deref>::Target)) {
  |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: struct `Data` is never constructed
 --> src/main.rs:3:8
  |
3 | struct Data {
  |        ^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: associated function `use_data` is never used
 --> src/main.rs:8:8
  |
8 |     fn use_data(&self, user: impl for <'a> FnOnce(<Box<&'a i32> as Deref>::Target)) {
  |        ^^^^^^^^

warning: deref which would be done by auto-deref
 --> src/main.rs:9:14
  |
9 |         user(*self.boxed)
  |              ^^^^^^^^^^^ help: try this: `self.boxed`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
  = note: requested on the command line with `-W clippy::explicit-auto-deref`

warning: `clpy` (bin "clpy" test) generated 3 warnings (run `cargo fix --bin "clpy" --tests` to apply 1 suggestion)
warning: `clpy` (bin "clpy") generated 3 warnings (3 duplicates)
    Finished dev [unoptimized + debuginfo] target(s) in 0.89
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (c5d82ed7a 2022-11-19)
binary: rustc
commit-hash: c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
commit-date: 2022-11-19
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
